src/Microsoft.AspNetCore.JsonPatch/Adapters/AdapterFactory.cs

```diff
@@ -13,20 +13,18 @@ namespace Microsoft.AspNetCore.JsonPatch.Adapters
     public class AdapterFactory : IAdapterFactory
     {
         /// <inheritdoc />
-        public virtual IAdapter Create(object target, IContractResolver contractResolver)
+        public virtual IAdapter Create(object target, JsonContract targetContract)
```

Comment:
Our discussion got me thinking: here I am arguing about `contractResolver` vs. `jsonContract` but none of my extension scenarios require either (I just use `target is` and an object type).  So now I have a different suggestion.  Let's pass `IAdapterFactory.Create` neither `contractResolver` nor `jsonContract`.  `Create`'s `target` parameter already contains all the information that method needs.  As you pointed out, we shouldn't preclude anyone from using a `contractResolver`.  But we're not!  For `AdapterFactory` for example, its constructor can accept `contractResolver`.

Reply:
You do bring up a good point, although that doesn't really follow the pattern established within the library and it would be possible to have a disconnect between the ContractResolver being used within the AdapterFactory and the rest of the library.  I would vote to leave it as is, or go back to passing in the ContractResolver to Create.  Nothing says that parameter needs to be used within the AdapterFactory.

Reply:
I'm not sure what you mean by "doesn't really follow the pattern established within the library and it would be possible to have a disconnect between the ContractResolver being used within the AdapterFactory and the rest of the library".  `ObjectVisitor` stores a reference to `_contractResolver`, and perhaps that's the case elsewhere in the library, too.

This new suggestion actually follows the same pattern that my app will use with `IAdapterFactory`.  That is, extra data that the factory needs will be passed into the factory's constructor.

Reply:
I simply mean that all the downstream classes (ObjectAdapter, ObjectVisitor, etc.) use the same ContractResolver that was initially passed into the JsonPatchDocument.  Having the AdapterFactory get its own instance out of that flow could cause a disconnect.  If the AdapterFactory has its own instance, then you could pass in a different ContractResolver to the JsonPatchDocument (and the classes it creates) than the one that the AdapterFactory is using.  I haven't through through the impacts of using a different resolver in the multiple areas.  This could certainly be an issue where the developer needs to do it "right" it just seems like we could insulate them a little bit from accidentally doing something wrong.  To your point, one could still pass in anything they want to a custom AdapterFactory, really we are arguing about the default implementation which the code should be able to make work however we want.

Reply:
I still don't follow.  The contract resolver passed to `JsonPatchDocument` is the one passed to `ObjectVisitor`'s constructor, which would be the one passed to  `AdapterFactory`'s constructor.

Reply:
Right, except the AdapterFactory is passed into the JsonPatchDocument, so you could create a JsonPatchDocument with contractResolverA passed in and pass in an instance of the AdapterFactory which uses contractResolverB.  Why you would is beside the point.  I am thinking about how to protect somebody new using JsonPatch that doesn't fully understand what is happening behind the scenes.

Reply:
I think we agree that a well-informed user wouldn't instantiate the default `AdapterFactory` class and pass that instance to `JsonPatchDocument`'s constructor, because there's no reason to.  The user would avoid the constructor that accepts an `IAdapterFactory`, and instead chose one that doesn't.   This results in the default `AdapterFactory` being instantiated on behalf of the user.

But even for new users... I don't know about you, but the first time I use the main class of a new API, I look for the constructor that has the fewest parameters.  If I need to configure it's behavior a certain way, then I'll look for a constructor overload that has the fewest parameters while still allowing me to do so.

The first time I used the `JsonPatchDocument.ApplyTo` overload that accepts an `ObjectAdapter`, I was caught off guard that the `ObjectAdapter`'s constructor accepts a `contractResolver`.  The first thing I did was to look for a `ContractResolver` property on my `JsonPatchDocument` instance, and indeed there's one there.  I expect a similar thought process from a user who unnecessarily wants to call the `JsonPatchDocument` constructor that accepts a `IAdapterFactory`.

Although I don't think it's necessary to protect a new user from mistakenly instantiating the default `AdapaterFactory` implementation, here's an idea for how to do it.  Make `AdapterFactory`'s constructor `internal`.  The user will see the compiler error and say, "I guess I'm not suppose to instantiate this class".  Before the user decides to implement `IAdapterFactory`, the user will likely consider the easier route and look for a `JsonPatchDocument` constructor overload that doesn't require it.

Reply:
I don't want to go the internal route.  I am actually overriding the base AdapterFactory in my implementation so I want to be able to instantiate it.  Let's pass the ContractResolver into the AdapterFactory as a constructor parameter.  The source code is available for a reason if somebody wants to know what is going on.  This still doesn't introduce any breaking changes and does keep things flexible.  I will work on that update tomorrow am and let you know when its ready for a review.  It shouldn't take long to make the change.

Reply:
Sounds good, thanks.  BTW, you could mark it as `protected internal` in that case.

Reply:
Now that I understand you want to derive from `AdapterFactory` and pass your subclass into `JsonPatchDocument`'s constructor, I see how the `AdapterFactory` constructor requiring a `contractResolver` would be slightly inconvenient.  You'd have to first instantiate `DefaultContractResolver` to pass to both your factory subclass's constructor and JsonPatchDocument`'s constructor.  In light of this, I defer to you on this one.  Take any of the routes we discussed, including passing `contractResolver` or `JsonContract` into `IAdapterFactory.Create` if you decide that's best.

Reply:
Ok, I am going to rollback to passing the contractResolver in IAdapterFactory.Create as I think it provides a good balance in flexibility while allowing the creation of the AdapterFactory to remain simple.  I will let you know when the code as been pushed.

Reply:
Updates made.  I think we should be good to go.